File: src/main/java/com/peasenet/gavui/Gui.java
/*
 * MIT License
 *
 * Copyright (c) 2022-2024, Gavin C. Pease
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.peasenet.gavui;

import com.peasenet.gavui.color.Color;
import com.peasenet.gavui.color.Colors;
import com.peasenet.gavui.math.BoxF;
import com.peasenet.gavui.math.PointF;
import com.peasenet.gavui.util.GuiUtil;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.font.TextRenderer;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.text.Text;

import java.util.ArrayList;
import java.util.UUID;

/**
 * @author gt3ch1
 * @version 2/28/2023
 * The base class for all gui elements.
 */
public class Gui {

    /**
     * The gui that was clicked.
     */
    protected static Gui clickedGui;
    /**
     * The original position of the gui.
     */
    protected final BoxF defaultPosition;
    /**
     * A randomly generated UUID for this gui.
     */
    private final UUID uuid = UUID.randomUUID();
    /**
     * The list of buttons(mods) in this dropdown.
     */
    protected ArrayList<Gui> children = new ArrayList<>();
    /**
     * The title of the gui.
     */
    protected Text title;
    protected String translationKey;
    /**
     * The symbol to be drawn to the left of the end of the box (like a checkbox, empty box, or arrow).
     */
    char symbol;
    /**
     * The offset used for the symbol (x).
     */
    int symbolOffsetX = -10;
    /**
     * The offset used for the symbol (y).
     */
    int symbolOffsetY = 1;
    private float transparency;
    private String settingKey;
    /**
     * Whether this element is a parent.
     */
    private boolean isParent = false;
    /**
     * The box that contains the gui.
     */
    private BoxF box;
    /**
     * The background color of the gui.
     */
    private Color backgroundColor = Colors.INDIGO;
    /**
     * Whether this gui is currently being dragged.
     */
    private boolean dragging;
    /**
     * Whether this gui is currently hidden.
     */
    private boolean hidden;
    /**
     * Whether this gui has been shrunk to fit a scrollbar.
     */
    private boolean shrunkForScroll = false;
    /**
     * Whether this gui is currently being hovered over.
     */
    private boolean hoverable = true;

    /**
     * Whether this GUI has a border to be drawn.
     */
    private boolean drawBorder = true;

    /**
     * Creates a new GUI menu.
     *
     * @param topLeft - The top left corner of the gui.
     * @param width   - The width of the gui.
     * @param height  - The height of the gui.
     * @param title   - The title of the gui.
     */
    public Gui(PointF topLeft, int width, int height, Text title) {
        box = new BoxF(topLeft, width, height);
        defaultPosition = BoxF.copy(box);
        this.title = title;
        dragging = false;
    }

    public Gui(GuiBuilder builder) {
        this.title = builder.getTitle();
        var w = builder.getWidth();
        if (title != null && MinecraftClient.getInstance().textRenderer != null)
            w = Math.max(w, MinecraftClient.getInstance().textRenderer.getWidth(title));

        box = new BoxF(builder.getTopLeft(), w, builder.getHeight());
        defaultPosition = BoxF.copy(box);
        dragging = false;
        this.translationKey = builder.getTranslationKey();
        setPosition(builder.getTopLeft());
        setBackground(builder.getBackgroundColor());
        setHoverable(builder.isHoverable());
        setSymbol(builder.getSymbol());
        setHidden(builder.isHidden());
        setHoverable(builder.isHoverable());
        setTransparency(builder.getTransparency());
        setDrawBorder(builder.getDrawBorder());
    }

    public static Gui getClickedGui() {
        return clickedGui;
    }

    public boolean getDrawBorder() {
        return drawBorder;
    }

    private void setDrawBorder(boolean drawBorder) {
        this.drawBorder = drawBorder;
    }

    public float getTransparency() {
        if (this.transparency == -1) {
            return GavUI.getAlpha();
        }
        return transparency;
    }

    /**
     * Sets the transparency of the gui. If set to -1, it will use the transparency of GavUI (set in the config).
     *
     * @param transparency - The transparency of the gui.
     */
    public void setTransparency(float transparency) {
        this.transparency = transparency;
        if (transparency == -1) {
            return;
        }
        if (transparency < 0) {
            this.transparency = 0f;
        }
        if (transparency > 1f) {
            this.transparency = 1f;
        }

    }

    public Text getTitle() {
        return title;
    }

    /**
     * Sets the title of the gui.
     *
     * @param title - The title of the gui.
     */
    public void setTitle(Text title) {
        this.title = title;
    }

    /**
     * Whether this gui should be brightened when hovered.
     *
     * @return Whether this gui should be brightened when hovered.
     */
    public boolean isHoverable() {
        return hoverable;
    }

    /**
     * Whether this gui should be brightened when hovered.
     *
     * @param hoverable Whether this gui should be brightened when hovered.
     */
    public void setHoverable(boolean hoverable) {
        this.hoverable = hoverable;
    }

    /**
     * Whether this element is a parent element.
     *
     * @return True if this element is a parent.
     */
    public boolean isParent() {
        return isParent;
    }

    /**
     * Sets whether this element is a parent.
     *
     * @param parent - True if we want this element to be a parent.
     */
    public void setParent(boolean parent) {
        isParent = parent;
    }

    /**
     * Sets the symbol that should be drawn on the right side of the gui.
     *
     * @param symbol - The symbol to draw.
     */
    public void setSymbol(char symbol) {
        this.symbol = symbol;
    }

    /**
     * Clears all children from this gui.
     */
    public void clearChildren() {
        children = new ArrayList<>();
    }

    /**
     * Gets the children of this gui.
     *
     * @return The children of this gui.
     */
    public ArrayList<Gui> getChildren() {
        return children;
    }

    /**
     * Adds an child element to this gui.
     *
     * @param child - The child element to add.
     */
    public void addElement(Gui child) {
        if (children.isEmpty()) {
            child.setPosition(new PointF(getX2() + 100, getY2() + 1));
            children.add(child);
            return;
        }
        // get last gui
        Gui lastButton = children.get(children.size() - 1);
        var lastY = lastButton.getY2();
        // set new gui position
        child.setPosition(new PointF(getX(), lastY + 2));
        children.add(child);
    }

    /**
     * Gets whether this gui is currently hidden.
     *
     * @return Whether this gui is currently hidden.
     */
    public boolean isHidden() {
        return hidden;
    }

    public void setHidden(boolean hidden) {
        if (hidden) {
            if (this.hasChildren()) {
                children.forEach(Gui::hide);
            }
        } else {
            if (this.hasChildren()) {
                children.forEach(Gui::show);
            }
        }
        this.hidden = hidden;
    }

    /**
     * Hides this gui.
     */
    public void hide() {
        setHidden(true);
    }

    /**
     * Shows this gui.
     */
    public void show() {
        setHidden(false);
    }

    /**
     * Sets the background color to the given color.
     *
     * @param color - The color to set the background to.
     */
    public void setBackground(Color color) {
        backgroundColor = color;
    }

    /**
     * Gets the x coordinate for the top left corner of the dropdown.
     *
     * @return The x coordinate for the top left corner of the dropdown.
     */
    public float getX() {
        return box.getX1();
    }

    /**
     * Gets the y coordinate for the top left corner of the dropdown.
     *
     * @return The y coordinate for the top left corner of the dropdown.
     */
    public float getY() {
        return box.getY1();
    }

    /**
     * Gets the x coordinate for the bottom right corner of the dropdown.
     *
     * @return The x coordinate for the bottom right corner of the dropdown.
     */
    public float getX2() {
        return box.getX2();
    }

    /**
     * Gets the y coordinate for the bottom right corner of the dropdown.
     *
     * @return The y coordinate for the bottom right corner of the dropdown.
     */
    public float getY2() {
        return box.getY2();
    }

    /**
     * Gets the width of the dropdown.
     *
     * @return The width of the dropdown.
     */
    public float getWidth() {
        return box.getWidth();
    }

    /**
     * Sets the width of the gui.
     *
     * @param width - The width of the gui.
     */
    public void setWidth(float width) {
        box = new BoxF(box.getTopLeft(), width, box.getHeight());
    }

    /**
     * Shrinks the gui to fit a scrollbar.
     *
     * @param parent - The parent gui.
     */
    public void shrinkForScrollbar(Gui parent) {
        if (shrunkForScroll && this.getWidth() == parent.getWidth()) return;
        if (this.getWidth() == parent.getWidth()) this.setWidth(getWidth() - 6);
        shrunkForScroll = true;
    }

    /**
     * Gets the height of the dropdown.
     *
     * @return The height of the dropdown.
     */
    public float getHeight() {
        return box.getHeight();
    }

    /**
     * Renders the clickable ui
     *
     * @param drawContext The draw context used to draw boxes on screen.
     * @param tr          The text render to use to draw text
     * @param mouseX      The x coordinate of the mouse.
     * @param mouseY      The y coordinate of the mouse.
     * @param delta       The change in time since the last render.
     */
    public void render(DrawContext drawContext, TextRenderer tr, int mouseX, int mouseY, float delta) {
        if (isHidden()) return;
        var matrixStack = drawContext.getMatrices();
        var bg = backgroundColor;
        if (bg == null)
            bg = GavUI.backgroundColor();
        if (mouseWithinGui(mouseX, mouseY) && hoverable)
            bg = bg.brighten(0.25f);
        GuiUtil.drawBox(bg, getBox(), matrixStack, getTransparency());
        var textColor = GavUI.textColor();
        if (title != null) {
            if (textColor.similarity(bg) < 0.3f) {
                textColor = textColor.invert();
                if (textColor.similarity(bg) < 0.3f)
                    textColor = Colors.WHITE;
            }
            drawText(drawContext, tr, title, getX() + 2, getY() + 1.5f, textColor);
        }
        drawSymbol(drawContext, tr, textColor);
        if (this.drawBorder)
            GuiUtil.drawOutline(GavUI.borderColor(), box, matrixStack);
        renderChildren(drawContext, tr, mouseX, mouseY, delta);
    }

    /**
     * Draws the GUI symbol.
     *
     * @param drawContext - The draw context to use.
     * @param tr          - The text renderer to use.
     */
    private void drawSymbol(DrawContext drawContext, TextRenderer tr, Color color) {
        if (symbol != '\0')
            drawText(drawContext, tr, Text.of(String.valueOf(symbol)), getX2() - 9f, getY() + 1.5f, color, false);
    }

    /**
     * Renders the children of this gui.
     *
     * @param drawContext - The draw context to use to draw the children.
     * @param tr          - The text renderer to use to draw the children.
     * @param mouseX      - The x coordinate of the mouse.
     * @param mouseY      - The y coordinate of the mouse.
     * @param delta       - The change in time since the last render.
     */
    private void renderChildren(DrawContext drawContext, TextRenderer tr, int mouseX, int mouseY, float delta) {
        if (!hasChildren()) return;
        for (Gui c : children)
            c.render(drawContext, tr, mouseX, mouseY, delta);
    }

    /**
     * Checks whether the mouse is clicked.
     *
     * @param mouseX - The x coordinate of the mouse.
     * @param mouseY - The y coordinate of the mouse.
     * @param button - The button that was clicked.
     * @return Whether the mouse was clicked.
     */
    public boolean mouseClicked(double mouseX, double mouseY, int button) {
        return false;
    }

    /**
     * Checks whether the mouse was dragged.
     *
     * @param mouseX - The x coordinate of the mouse.
     * @param mouseY - The y coordinate of the mouse.
     * @param button - The button that was dragged.
     * @param deltaX - The change in x coordinate.
     * @param deltaY - The change in y coordinate.
     * @return Whether the mouse was dragged.
     */
    public boolean mouseDragged(double mouseX, double mouseY, int button, double deltaX, double deltaY) {
        return false;
    }

    /**
     * Checks whether the mouse was scrolled
     *
     * @param mouseX - The x coordinate of the mouse.
     * @param mouseY - The y coordinate of the mouse.
     * @param amount - The amount of scroll.
     * @return Whether the mouse was scrolled.
     */
    public boolean mouseScrolled(double mouseX, double mouseY, double amount) {
        return false;
    }

    /**
     * Whether the current window is being dragged.`
     *
     * @return True if the current window is being dragged.
     */
    public boolean isDragging() {
        return dragging;
    }

    /**
     * Sets whether the current window is being dragged.
     *
     * @param dragging - Whether the current window is being dragged.
     */
    public void setDragging(boolean dragging) {
        this.dragging = dragging;
        for (Gui child : children)
            child.setDragging(dragging);
        clickedGui = null;
    }

    public BoxF getBox() {
        return box;
    }

    /**
     * Resets the position to the default position.
     */
    public void resetPosition() {
        box = BoxF.copy(defaultPosition);
    }

    /**
     * Gets the current location of the top left corner of the gui.
     *
     * @return The current location of the top left corner of the gui.
     */
    public PointF getPosition() {
        return box.getTopLeft();
    }

    /**
     * Sets the top left corner of the gui element to the given point.
     *
     * @param position - The point to set the top left corner of the gui element to.
     */
    public void setPosition(PointF position) {
        box.setTopLeft(position);
    }

    /**
     * Sets the middle of the gui element to the given point.
     *
     * @param position - The point to set the middle of the gui element to.
     */
    public void setMidPoint(PointF position) {
        box.setMiddle(position);
    }

    /**
     * Gets whether the mouse coordinates are within the bounds of the gui.
     *
     * @param mouseX - The x coordinate of the mouse.
     * @param mouseY - The y coordinate of the mouse.
     * @return Whether the mouse coordinates are within the bounds of the gui.
     */
    public boolean mouseWithinGui(double mouseX, double mouseY) {
        return mouseX >= getX() && mouseX <= getX2() && mouseY >= getY() && mouseY <= getY2() && !isHidden();
    }

    /**
     * Gets the background color of the gui.
     *
     * @return The background color of the gui.
     */
    public Color getBackgroundColor() {
        return backgroundColor;
    }

    /**
     * Whether this element has children.
     *
     * @return True if this element has children.
     */
    public boolean hasChildren() {
        return !children.isEmpty();
    }

    public void setShrunkForScrollbar(boolean b) {
        shrunkForScroll = b;
        if (hasChildren()) for (Gui c : children)
            c.setShrunkForScrollbar(b);
    }

    public UUID getUUID() {
        return uuid;
    }

    @Override
    public boolean equals(Object o) {
        if (o == null) return false;
        if (o instanceof Gui g) {
            return g.getUUID().equals(uuid);
        }
        return false;
    }


    /**
     * Sets the default position of the gui.
     *
     * @param newDefaultPosition - The new default position of the gui.
     */
    public void setDefaultPosition(BoxF newDefaultPosition) {
        defaultPosition.from(newDefaultPosition);
    }

    //NOTE: I wrote this so that I don't have to deal with the API in drawContext and I can just use the textRenderer directly.
    protected void drawText(DrawContext drawContext, TextRenderer textRenderer, Text text, float x, float y, Color color, boolean shadow) {
        drawContext.drawText(textRenderer, text.asOrderedText(), (int)x, (int)y, color.getAsInt(), shadow);
    }

    protected void drawText(DrawContext drawContext, TextRenderer textRenderer, String text, float x, float y, Color color) {
        drawText(drawContext, textRenderer, Text.of(text), x, y, color, false);
    }

    protected void drawText(DrawContext drawContext, TextRenderer textRenderer, Text text, float x, float y, Color color) {
        drawText(drawContext, textRenderer, text, x, y, color, false);
    }

    public String getTranslationKey() {
        return this.translationKey;
    }
}

File: src/main/kotlin/com/peasenet/gavui/Gui.kt
/*
 * MIT License
 *
 * Copyright (c) 2022-2025, Gavin C. Pease
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.peasenet.gavui

import com.peasenet.gavui.color.Color
import com.peasenet.gavui.color.Colors
import com.peasenet.gavui.math.BoxF
import com.peasenet.gavui.math.PointF
import com.peasenet.gavui.util.GuiUtil
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Font
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.network.chat.Component
import java.util.*
import java.util.function.Consumer
import kotlin.math.max

/**
 * The base class for all gui elements.a
 * @param builder The builder for this gui.
 * @param children The children of this gui.
 * @param title The title of this gui.
 * @param translationKey The translation key of this gui.
 * @param symbol The symbol of this gui.
 * @param isParent Whether this gui is a parent.
 * @param box The box of this gui.
 * @param backgroundColor The background color of this gui.
 * @param canHover Whether this gui can be hovered over.
 * @param drawBorder Whether this gui should draw a border.
 * @param height The height of this gui.
 *
 * @author GT3CH1
 * @version 11-27-2025
 * @since 2/28/2023
 */
open class Gui(
    builder: GuiBuilder<*>,
    var children: ArrayList<Gui> = ArrayList(),
    var title: Component? = builder.title,
    var translationKey: String? = builder.translationKey,
    var symbol: String? = builder.symbol,
    var isParent: Boolean = builder.isParent,
    var box: BoxF = BoxF(builder.topLeft, builder.width, builder.height),
    var backgroundColor: Color = builder.backgroundColor,
    var canHover: Boolean = builder.canHover,
    var drawBorder: Boolean = builder.drawBorder,
    var height: Float = builder.height,
) {
    /**
     * The original position of the gui.
     */
    private val defaultPosition: BoxF

    /**
     * A randomly generated UUID for this gui.
     */
    val uUID: UUID = UUID.randomUUID()

    /**
     * The offset for the symbol.
     */
    var symbolOffsetX: Int = -10

    /**
     * The offset for the symbol.
     */
    var symbolOffsetY: Int = 1

    /**
     * How transparent the gui is.
     */
    private var transparency: Float = -1f
        get() {
            if (field == -1f) {
                return GavUI.alpha
            }
            return field
        }
        set(transparency) {
            if (transparency == -1f) {
                return
            }
            field = transparency.coerceIn(0f, 1f)
        }

    /**
     * Whether this gui is being dragged.
     */
    var dragging: Boolean = false
        set(value) {
            field = value
            for (child in children) child.dragging = (value)
        }

    /**
     * Whether this gui is hidden.
     */
    var isHidden: Boolean = false
        set(hidden) {
            field = hidden
            if (this.hasChildren()) {
                children.forEach(Consumer { obj: Gui -> obj.isHidden = hidden })
            }
        }

    /**
     * Whether this gui has been shrunk for a scrollbar.
     */
    private var shrunkForScroll = false
        set(value) {
            field = value
            if (hasChildren()) for (c in children) c.shrunkForScroll = value
        }

    /**
     * Clears all children from this gui.
     */
    fun clearChildren() {
        children.clear()
    }

    /**
     * Adds an child element to this gui.
     *
     * @param child - The child element to add.
     */
    open fun addElement(child: Gui) {
        if (children.isEmpty()) {
            child.position = PointF(x2 + 100, y2 + 1)
            children.add(child)
            return
        }
        // get last gui
        val lastButton = children[children.size - 1]
        val lastY = lastButton.y2
        // set new gui position
        child.position = PointF(x, lastY + 2)
        children.add(child)
    }

    /**
     * Hides this gui.
     */
    fun hide() {
        isHidden = true
    }

    /**
     * Shows this gui.
     */
    fun show() {
        isHidden = false
    }


    val x: Float
        get() = box.x1

    val y: Float
        get() = box.y1

    val x2: Float
        get() = box.x2

    val y2: Float
        get() = box.y2

    var width: Float
        get() = box.width
        set(width) {
            box = BoxF(box.topLeft, width, box.height)
        }

    /**
     * Shrinks the gui to fit a scrollbar.
     *
     * @param parent - The parent gui.
     */
    fun shrinkForScrollbar(parent: Gui) {
        if (shrunkForScroll) return
        width -= 6
        shrunkForScroll = true
    }

    /**
     * Renders the clickable ui
     *
     * @param drawContext The draw context used to draw boxes on screen.
     * @param tr          The text render to use to draw text
     * @param mouseX      The x coordinate of the mouse.
     * @param mouseY      The y coordinate of the mouse.
     * @param delta       The change in time since the last render.
     */
    open fun render(
        drawContext: GuiGraphics,
        tr: Font,
        mouseX: Int,
        mouseY: Int,
        delta: Float,
    ) {
        if (isHidden) return
        var bg = backgroundColor
        if (mouseWithinGui(mouseX, mouseY) && canHover) bg = bg.brighten(0.25f)
        bg = bg.withAlpha(transparency)
        drawContext.guiRenderState.up()
        GuiUtil.fill(box, drawContext, bg)
        drawContext.guiRenderState.up()
        GuiUtil.drawOutline(box, drawContext, GavUI.borderColor().withAlpha(GavUI.alpha))
        var textColor = GavUI.textColor()
        if (title != null) {
            if (textColor.similarity(bg) < 0.3f) {
                textColor = textColor.invert()
                if (textColor.similarity(bg) < 0.3f) textColor = Colors.WHITE
            }
            val titleH = tr.lineHeight
            val offsetY = (box.height - titleH)
            drawText(drawContext, tr, title!!, x.toInt() + 2, y.toInt() + offsetY.toInt(), textColor)
        }
        drawSymbol(drawContext, tr, textColor, 0f, 2f)

        if (this.drawBorder) GuiUtil.drawOutline(GavUI.borderColor(), box, drawContext.pose())
        renderChildren(drawContext, tr, mouseX, mouseY, delta)
    }

    /**
     * Draws the GUI symbol.
     *
     * @param drawContext - The draw context to use.
     * @param tr          - The text renderer to use.
     */
    protected fun drawSymbol(
        drawContext: GuiGraphics,
        tr: Font,
        color: Color,
        offsetX: Float = 0f,
        offsetY: Float = 0f,
    ) {
        if (symbol == null) return
        val symbolWidth = tr.width(symbol.toString())
        drawContext.drawString(
            tr,
            symbol,
            (x2 - symbolWidth + offsetX).toInt(),
            (y + symbolOffsetY + offsetY).toInt(),
            color.asInt,
            false,
        )
    }

    /**
     * Renders the children of this gui.
     *
     * @param drawContext - The draw context to use to draw the children.
     * @param tr          - The text renderer to use to draw the children.
     * @param mouseX      - The x coordinate of the mouse.
     * @param mouseY      - The y coordinate of the mouse.
     * @param delta       - The change in time since the last render.
     */
    private fun renderChildren(
        drawContext: GuiGraphics,
        tr: Font,
        mouseX: Int,
        mouseY: Int,
        delta: Float,
    ) {
        if (!hasChildren()) return
        for (c in children) c.render(drawContext, tr, mouseX, mouseY, delta)
    }

    /**
     * Checks whether the mouse is clicked.
     *
     * @param mouseX - The x coordinate of the mouse.
     * @param mouseY - The y coordinate of the mouse.
     * @return Whether the mouse was clicked.
     */
    open fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        return false
    }

    /**
     * Checks whether the mouse was dragged.
     *
     * @param mouseX - The x coordinate of the mouse.
     * @param mouseY - The y coordinate of the mouse.
     * @param button - The button that was dragged.
     * @param deltaX - The change in x coordinate.
     * @param deltaY - The change in y coordinate.
     * @return Whether the mouse was dragged.
     */
    open fun mouseDragged(mouseX: Double, mouseY: Double, button: Int, deltaX: Double, deltaY: Double): Boolean {
        return false
    }

    /**
     * Checks whether the mouse was scrolled
     *
     * @param mouseX - The x coordinate of the mouse.
     * @param mouseY - The y coordinate of the mouse.
     * @param amount - The amount of scroll.
     * @return Whether the mouse was scrolled.
     */
    open fun mouseScrolled(mouseX: Double, mouseY: Double, amount: Double): Boolean {
        return false
    }

    /**
     * Whether the current window is being dragged.`
     *
     * @return True if the current window is being dragged.
     */
    @Deprecated("Use dragging instead.", ReplaceWith("dragging"))
    fun isDragging(): Boolean {
        return dragging
    }

    /**
     * Resets the position to the default position.
     */
    open fun resetPosition() {
        box = BoxF.copy(defaultPosition)
    }

    var position: PointF
        get() = box.topLeft
        set(position) {
            box.topLeft = position
        }

    /**
     * Sets the middle of the gui element to the given point.
     *
     * @param position - The point to set the middle of the gui element to.
     */
    fun setMidPoint(position: PointF) {
        box.setMiddle(position)
    }

    /**
     * Gets whether the mouse coordinates are within the bounds of the gui.
     *
     * @param mouseX - The x coordinate of the mouse.
     * @param mouseY - The y coordinate of the mouse.
     * @return Whether the mouse coordinates are within the bounds of the gui.
     */
    open fun mouseWithinGui(mouseX: Double, mouseY: Double): Boolean {
        return mouseX in x..x2 && mouseY >= y && mouseY <= y2 && !isHidden
    }

    fun mouseWithinGui(mouseX: Int, mouseY: Int): Boolean {
        return mouseWithinGui(mouseX.toDouble(), mouseY.toDouble())
    }

    /**
     * Whether this element has children.
     *
     * @return True if this element has children.
     */
    fun hasChildren(): Boolean {
        return children.isNotEmpty()
    }

    fun setShrunkForScrollbar(b: Boolean) {
        shrunkForScroll = b
        if (hasChildren()) for (c in children) c.setShrunkForScrollbar(b)
    }

    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (other is Gui) {
            return other.uUID == uUID
        }
        return false
    }


    /**
     * Sets the default position of the gui.
     *
     * @param newDefaultPosition - The new default position of the gui.
     */
    fun setDefaultPosition(newDefaultPosition: BoxF) {
        defaultPosition.from(newDefaultPosition)
    }

    //NOTE: I wrote this so that I don't have to deal with the API in drawContext and I can just use the textRenderer directly.
    protected fun drawText(
        drawContext: GuiGraphics,
        textRenderer: Font,
        text: Component,
        x: Int,
        y: Int,
        color: Color,
        shadow: Boolean = false,
    ) {
        drawContext.drawString(textRenderer, text, x, y, color.asInt, shadow)
    }

    override fun hashCode(): Int {
        return uUID.hashCode()
    }

    companion object {
        /**
         * The gui that was clicked.
         */
        @JvmStatic
        var clickedGui: Gui? = null
    }

    init {
        this.title = builder.title
        var w = builder.width
        if (title != null && Minecraft.getInstance().font != null) w = max(
            w.toDouble(),
            Minecraft.getInstance().font.width(title).toDouble()
        ).toFloat()
        box = BoxF(builder.topLeft, w, builder.height)
        defaultPosition = BoxF.copy(box)
        dragging = false
        this.translationKey = builder.translationKey
        position = builder.topLeft
        backgroundColor = builder.backgroundColor
        symbol = (builder.symbol)
        isHidden = builder.isHidden
        canHover = builder.canHover
        transparency = builder.transparency
        drawBorder = builder.drawBorder
        isParent = builder.isParent
    }
}
